### ISSUE TITLE
What does autosklearn validation score refer to?

### ISSUE THREAD

**Issue by iXanthos** [automl/auto-sklearn#837]
Hello, it is not clear to me what the "best validation score" is depicting when I use the `sprint_statistics()` function. Is it the validation score of the ensembling or just of one of the models? I used to think it is the score of the ensemble, but after discussions with other users, I am not that confident anymore. It is critical to have a train estimation using the final model (the ensemble), so if this is not it, then I would also like to ask where I can find this information, once the fitting of autosk is completed.

Thanks in advance.

**Comment by mfeurer** [comment]
> Is it the validation score of the ensembling or just of one of the models?

It's the score of the best single model.

> It is critical to have a train estimation using the final model (the ensemble)

I assume you mean an estimate of the performance on the internal validation set the ensemble is constructed on, right? Apparently, we don't store that information and you'd have to retrieve it from the log files. I'd be happy about a pull request fixing this issue.

**Comment by iXanthos** [comment]
> > Is it the validation score of the ensembling or just of one of the models?
> 
> It's the score of the best single model.
> 
> > It is critical to have a train estimation using the final model (the ensemble)
> 
> I assume you mean an estimate of the performance on the internal validation set the ensemble is constructed on, right? Apparently, we don't store that information and you'd have to retrieve it from the log files. I'd be happy about a pull request fixing this issue.
 
Yes, this is what I mean. Since this might take a while, where can I find this information in the log files?

**Comment by mfeurer** [comment]
It's the final entries of lines like:
```
Ensemble Selection:
	Trajectory: 0: 0.015233 1: 0.009615 2: 0.009615 3: 0.009615 4: 0.009615 5: 0.009615 6: 0.009615 7: 0.009615 8: 0.009615 9: 0.009615 10: 0.009615 11: 0.009615 12: 0.009615 13: 0.009615 14: 0.009615 15: 0.009615 16: 0.009615 17: 0.009615 18: 0.009615 19: 0.009615 20: 0.009615 21: 0.009615 22: 0.009615 23: 0.009615 24: 0.009615 25: 0.009615 26: 0.009615 27: 0.009615 28: 0.009615 29: 0.009615 30: 0.009615 31: 0.009615 32: 0.009615 33: 0.009615 34: 0.009615 35: 0.009615 36: 0.009615 37: 0.009615 38: 0.009615 39: 0.009615 40: 0.009615 41: 0.009615 42: 0.009615 43: 0.009615 44: 0.009615 45: 0.009615 46: 0.009615 47: 0.009615 48: 0.009615 49: 0.009615
```

**Comment by iXanthos** [comment]
> It's the final entries of lines like:
> 
> ```
> Ensemble Selection:
> 	Trajectory: 0: 0.015233 1: 0.009615 2: 0.009615 3: 0.009615 4: 0.009615 5: 0.009615 6: 0.009615 7: 0.009615 8: 0.009615 9: 0.009615 10: 0.009615 11: 0.009615 12: 0.009615 13: 0.009615 14: 0.009615 15: 0.009615 16: 0.009615 17: 0.009615 18: 0.009615 19: 0.009615 20: 0.009615 21: 0.009615 22: 0.009615 23: 0.009615 24: 0.009615 25: 0.009615 26: 0.009615 27: 0.009615 28: 0.009615 29: 0.009615 30: 0.009615 31: 0.009615 32: 0.009615 33: 0.009615 34: 0.009615 35: 0.009615 36: 0.009615 37: 0.009615 38: 0.009615 39: 0.009615 40: 0.009615 41: 0.009615 42: 0.009615 43: 0.009615 44: 0.009615 45: 0.009615 46: 0.009615 47: 0.009615 48: 0.009615 49: 0.009615
> ```

I am sorry, but this is unclear for me. In what files do I find this information and lets say specifically for classification, which value depicts the estimated performance?

Thanks in advance.

**Comment by mfeurer** [comment]
> In what files do I find this information

In the general Auto-sklearn log file, which is in the `tmp_folder`.

> which value depicts the estimated performance

This is the trajectory of ensemble selection. If you take a look into the [original paper](https://www.cs.cornell.edu/~alexn/papers/shotgun.icml04.revised.rev2.pdf), it is the green line in Figure 1. Auto-sklearn by default builds an ensemble of size 50. The last entry is the training loss of an ensemble of size 50 (and the other entries are for smaller ensemble sizes).

**Comment by iXanthos** [comment]
I see. So the loss is always RMS error, regardless of the task at hand (classification, regression)? Sorry for the multiple questions, I just need to be sure.

**Comment by mfeurer** [comment]
> Sorry for the multiple questions, I just need to be sure.

No worries, I'm doing my best to answer.

> So the loss is always RMS error, regardless of the task at hand (classification, regression)?

No, it's actually the metric you set in `AutosklearnClassifier.fit()`. In case that is a metric to be maximized, this is 1-metric. In the case of accuracy, this is the error rate.

**Comment by iXanthos** [comment]
This is super useful, thanks. Still, I have trouble finding the file containing this information. When I look at the output_folder, it is empty. This is the command I run:
`
automl = autosklearn.classification.AutoSklearnClassifier(
        time_left_for_this_task=time_threshold,
        per_run_time_limit=int(time_threshold / 10),  # change this division
        # custom folder (/tmp/dataset_name)
        tmp_folder=tmp_folder + '/' + dataset_info_dict['dataset_name'] + "_tmp",
        output_folder=tmp_folder + '/' + dataset_info_dict['dataset_name'] + "_out",
        n_jobs=8,
        ml_memory_limit = 25000,
        ensemble_memory_limit = 15000
    )`

 should I do something else to keep the out logs, or is this info located elsewhere?

**Comment by iXanthos** [comment]
> > In what files do I find this information
> 
> In the general Auto-sklearn log file, which is in the `tmp_folder`.
Ok, there are some issues I come up against when trying to retrieve this information. First of all, I use `n_jobs = 8`, as I run these experiments on a multicore setup. This translates to 8 different .log files in the `_tmp` folder of each analysis. I manually searched each of them and found out only 1 of them contains the loss information, which is good. 
Still, when I open the correct log and search for `Ensemble selection`, I find this entry multiple times in the same .log file, each time with different results. Which one is the correct one?

**Comment by mfeurer** [comment]
Hey, it's in the `tmp_folder`, not the `output_folder`. 

Another note, the `ml_memory_limit` is per job. I assume that you have 25 GB in total and not 200GB, so a good limit here would be 3 or 4GB.

**Comment by iXanthos** [comment]
> Hey, it's in the `tmp_folder`, not the `output_folder`.

Yes, I saw that later, hence why removed my comment. I have updated the comment now.

> Another note, the `ml_memory_limit` is per job. I assume that you have 25 GB in total and not 200GB, so a good limit here would be 3 or 4GB.
Ah, I see. About the memory limits. I am using 8 jobs in a machine with 50GB of memory. What are some suggested memory limits for the ml_memory_limit and ensemble_memory_limit ?

**Comment by mfeurer** [comment]
Yes, Ensemble Selection is only started for one of the processes. It is run as a background process and builds a new ensemble whenever new models to ensemble are available. This means that you get a trajectory of ensembles (and each one has a trajectory itself). The last of these entries is the one you're looking for.

If you have 50GB of RAM and 8 cores that you can devote completely to Auto-sklearn, I would (from the top of my head) go with 5GB `ml_memory_limit` and 5GB `ensemble_memory_limit`. This leaves 5GB for Bayesian optimization and the OS.

**Comment by iXanthos** [comment]
Thanks for the discussion and clarifications. I will close this thread as my questions have been answered. If I come across another issue in the same context, I will comment here.

Regards,
Iordanis
